### PR TITLE
feat(core): add support for tags with `(print-)affected(:*)` and `run-many`

### LINKED ISSUE #1621: Exclude Apps in affected scripts by tags from nx.json
Currently the only way to exclude apps and libs from being processed by the affected scripts, is by their names. 
It would be useful, if we could exclude apps and libs in the affected scripts by their tags which can be defined in nx.json

For example having the following nx.json
```
....
    "booking": {
      "tags": ["booking"]
    },
    "error-page": {
      "tags": ["error-page"]
    },
    "my-nanosite": {
      "tags": ["my-nanosite", "nanosites"]
    },
    "my-other-nanosite": {
      "tags": ["my-other-nanosite", "nanosites"]
    },
...
```
and running `npm run affected:apps -- --base=origin/develop --exclude="tags:nanosites"` or something similar :) only apps should printed out, which are affected by changes and do not have the `nanosites` tag.
It should also be possible to pass a list of tags to be excluded.
This would save us a lot of headaches, because we have more than 50 nanosites and adding them by name in the exclude part is tedious and error prone.

I know the nx.json is used for linting, but could that purpose be extended?

### LINKED ISSUE #2675: [FEAT] run-many --tags=tag1,tag2
Would be nice if we could use tags with the `nx run-many` command.

Example
```shell
> nx run-many --target=build --tags=core,demo-app
```

### PR DESCRIPTION
**Pull request by fguitton**
⚠️ _Edit: The behaviour of this pull request has changed to address feedbacks. See https://github.com/nrwl/nx/pull/10085#issuecomment-1493061063 for more information_

## Current Behavior
When running the `(print-)affected(:*)` command you get all affected projects for a target.
With the `run-many` command filtering is only possible with `--projects`.
Both accept a `--all`, it being warned for in `affected`.

## Expected Behavior
We would expect to be able to select projects by `tags`.

## This PR
This could be seen as a first step only acting to select projects in the same way `--projects` does it for `run-many`. Tags are not applied in compound but rather as `OR` statement in comma separated fashion.

That is in the example below the command would select projects that either contain the tag `sdk` or the tag `app`, not necessarily projects containing both tags.
```
nx affected --tags=sdk,app --target=test
```

## Related Issue(s)
This relates to very long standing #1621, #2675 or #8292 and a Nx **v14** compatible treatment while being different from #8364 which attempts to support globs but does not address `run-many`.

Closes #1621
Closes #2675

### CODE REVIEW

**Comment by fguitton** [comment]
Hello @vsavkin !
I know you had interest in looking at this over the past few years.
It is the second PR opened for this feature but it addressed things slightly differently.
I am happy to bring any changes required, any feedback/direction is most welcome
We have a lot of need for this in our company. 

Thank you for your help and valuable time 🙏🏼 !

File: packages/nx/src/utils/project-graph-utils.ts [hunk 1]
```diff
@@ -29,6 +29,14 @@ export function projectHasTargetAndConfiguration(
   );
 }
 
+export function projectHasTag(project: ProjectGraphProjectNode, tag: string) {
+  return !!(
+    project.data &&
+    project.data.tags &&
+    project.data.tags.includes(tag)
+  );
```
Comment (gioragutt) [inline packages/nx/src/utils/project-graph-utils.ts:37]:
```suggestion
  return !!project?.data?.tags?.includes(tag);
```

Comment (fguitton) [reply packages/nx/src/utils/project-graph-utils.ts:37]: Happy to bring the change ! This was done only to make the code style the same as for `projectHasTarget` and `projectHasTargetAndConfiguration` in the same file 😅

Comment (gioragutt) [reply packages/nx/src/utils/project-graph-utils.ts:37]: I don't think it's a problem if you'll even fix the existing code, it's only for the better :)

Comment (fguitton) [reply packages/nx/src/utils/project-graph-utils.ts]: That's now done in 9dafbd4, for other other methods too 😉

**Comment by gioragutt** [comment]
Any update on this? I would love to use it 😁 Also, I thought it might be worth adding this to `print-affected` as well

**Comment by fguitton** [comment]
> I thought it might be worth adding this to `print-affected` as well

Hello ! The `print-affected` command should already work out of the box with the changes proposed at it is considered a special branch of the `affected` command and the project filtering gets done upstream of that.

https://github.com/nrwl/nx/blob/7f5137da84e5d0f1b31f9011c57fc0e3bfaa9252/packages/nx/src/command-line/affected.ts#L37

https://github.com/nrwl/nx/blob/7f5137da84e5d0f1b31f9011c57fc0e3bfaa9252/packages/nx/src/command-line/affected.ts#L79-L113

I can imagine there is latent question with regard to the use/semantic of `--exclude` which only applies to project names. I do not necessarily have a set opinion on this. In effect a tag filtering can very much _replace_ the use of `--exclude`.

**Comment by yharaskrik** [comment]
Would love to see this merged into Nx and my PR closed.

**Comment by epechuzal** [comment]
This would be a big QoL feature for us, would love to see this in the next release :D

**Comment by sylvainar** [comment]
Can't wait for this to be merged :smile:

**Comment by josh-stevens** [comment]
pls

**Comment by starch** [comment]
This would be a great QoL feature for our team 👍

**Comment by jon9090** [comment]
When this PR will be merged?

**Comment by jon9090** [comment]
@gioragutt

**Comment by gioragutt** [comment]
> @gioragutt 

Can you add some context to why you tagged me? 😅

**Comment by ianldgs** [comment]
Would really like this!

Use-case: only publish to the public npm registry packages tagged with `open-source`.

**Comment by mrfelton** [comment]
Also interested in this. I want to be able to get a list of affected projects filtered by tag. Use case is for use with the nx dotnet  plugin. When I generate a dotnet app it actually generates 2 app. One is the main dotnet app. The other is a test app. I want to get a list of affected apps excluding the test apps.

**Comment by msz13** [comment]
When this feature will be published? I'm wating for it. Use case: create separate ci/cd pipelines for projects in different technologies, like js and dotnet (with nx dotnet plugin)

**Comment by gioragutt** [comment]
@fguitton thank you for keeping this up to date for when they finally merge this 🙏🏻 I'm amused every time I get a notification about a rebase 😂

**Comment by fguitton** [comment]
> I'm amused every time I get a notification about a rebase

Haha ! You're most welcome. I guess I'm being a bit overzealous on this one 😅 .
It's almost going to be the oldest non-merged (or closed) pull-request now ! 😄

**Comment by mrfelton** [comment]
Is there anything specific that is blocking this from being merged?

**Comment by fguitton** [comment]
@gioragutt 
> for when they finally merge this

We are now officially the older open pull request at sea. Perhaps this means we're next 😃

**Comment by gioragutt** [comment]
> @gioragutt 
> 
> > for when they finally merge this
> 
> 
> 
> We are now officially the older open pull request at sea. Hehe. Perhaps this means we're next ! 😃 

That's actually funny 😄 and again, kudos to you for keeping it up all this time 🙏🏻

@

File: docs/generated/packages/nx.json [hunk 2]
```diff
@@ -0,0 +1,345 @@
+{
```
Comment (AgentEnder) [inline docs/generated/packages/nx.json:1]: I'm not sure this file should have been added. Can you `rm -rf docs && yarn documentation` on your next rebase?

**Review by AgentEnder — CHANGES_REQUESTED** [review]
Overall this implementation looks good to me. I'll discuss more with @FrozenPandaz later, but on your next rebase can you double check the docs changes?

Also, I don't see any e2e tests added for this. If you could add a few (1 for affected, 1 for run-many) it would be really helpful.

I'm sorry this PR has kind of got left behind for a while, we will try to do better in the future for this.

File: packages/nx/src/command-line/affected.ts [hunk 3]
```diff
@@ -162,6 +166,39 @@ function projectsToRun(
...
+        ...nodes,
+        [project.name]: project,
+      }),
+      {}
+    ),
+    dependencies: affectedGraph.dependencies,
+  } as ProjectGraph;
+  return res;
+}
+
```
Comment (AgentEnder) [inline packages/nx/src/command-line/affected.ts:201]:
```suggestion
function affectedProjectsWithTags(
  affectedGraph: ProjectGraph,
  nxArgs: NxArgs
): ProjectGraph {
  const builder = new ProjectGraphBuilder(affectedGraph)
  const affectedProjectNodes = Object.entries(affectedGraph.nodes);

  for (const [nodeName, projectNode] of affectedProjectNodes) {
    if ( !hasMatchingTags(projectNode, nxArgs.tags ) {
      builder.removeNode(nodeName)
    }
  }
  
  return builder.getUpdatedProjectGraph();
}

```

**Review by AgentEnder — CHANGES_REQUESTED** [review]
This can also be a bit simpler if we use the project graph builder.

NOTE - `hasMatchingTags` doesn't exist, you'll need to implement it. Most of the functionality should come from findMatchingProjects.

**Comment by AgentEnder** [comment]
@fguitton Actually, after discussing this with @vsavkin a bit more today there is a bit more involved change that we'd like to make to properly support this and open some more doors in the future. Have you joined our community slack? I'd love to connect there and offer some more insight. 

Here's the link if you aren't there yet, just dm me 😄: https://go.nrwl.io/join-slack

**Comment by fguitton** [comment]
@AgentEnder,
So I found a bit of time to explore the direction you indicated and I have updated the PR to use this new approach.

The `findMatchingProjects` function has been modified to a operated on the following pattern: `[tag:]filter[,...]`.
I have also removed the `Set(...)` performance bailout but making sure the performance test on CI were satisfied.

The principle stays the same and the change still covers `(print-)affected(:*)` and `run-many` but everything now happens through `--projects` argument. I have also made sure to permit a similar logic for the `--exclude` commutator as well, making sure is it backward compatible.

Negative filtering is supported and the syntax is flexible; either `!tag:ui` or `tag:!ui` indifferently.

Mix of project name and tag filters are also supported with an _non-chained_, negation-takes-all_ behaviour. Meaning the order of filters have no impact and we prefer to exclude in all cases.

I took the liberty to remove the mentions that tags only served for linting as well as added a couple of unit tests to insure that things behaved properly.

Feedback is obviously welcome, looking forward to this being merged in !

Best wishes 🌴

**Comment by AgentEnder** [comment]
Awesome! Thanks for the updates @fguitton. I'll look over this either today or tomorrow.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by AgentEnder — APPROVED** [review]

File: packages/nx/src/utils/find-matching-projects.ts [hunk 4]
```diff
@@ -1,46 +1,102 @@
...
 
-    matchedProjectNames.forEach((matchedProjectName) => {
-      if (exclude) {
-        excludedProjects.add(matchedProjectName);
-      } else {
-        selectedProjects.add(matchedProjectName);
+        if (minimatch.match(tags, pattern.value).length)
+          (pattern.exclude ? excludedProjects : selectedProjects).add(
+            projectName
+          );
```
[the remark this answers is not in the document]
Comment (AgentEnder) [reply packages/nx/src/utils/find-matching-projects.ts:76]: Yeah, I'll look at this in the future

**Review by FrozenPandaz — APPROVED** [review]

**Review by bcabanes — APPROVED** [review]

**Comment by fguitton** [comment]
Thank you @AgentEnder ! Extending to the whole Nx team of course ! Thrilled to see this merged 🎉

**Comment by gioragutt** [comment]
Congrats @fguitton, you've put a lot of work on this and it shows! Thank you for all your effort, I'm sure it'll bring immense value to a lot of people 🙏🏻🙏🏻🙏🏻

**Pull request merged**